### PR TITLE
Support for mastery acquisition specification

### PR DESCRIPTION
**Pull request by arikpamnani**
## How did you test this change?

- [x] Test(s) added
- [x] Manual testing (please provide screenshots/recordings)
- [ ] No testing (please provide an explanation)

### CODE REVIEW

**Comment by gayathrir11** [comment]
I think you need to get CLA coverage. Otherwise we won't be able to merge this change

File: packages/legend-extension-dsl-mastery/src/graph-manager/DSL_Mastery_GraphManagerHelper.ts [hunk 1]
```diff
@@ -36,3 +42,76 @@ export const getOwnMasterRecordDefinition = (
...
+    DataProvider,
+    `Can't find master record definition '${path}'`,
+  );
+
+export const getOwnDataProvider = (
+  path: string,
+  graph: BasicModel,
+): DataProvider =>
+  guaranteeNonNullable(
+    graph.getOwnNullableExtensionElement(path, DataProvider),
```
[the remark this answers is not in the document]
Comment (arikpamnani) [reply packages/legend-extension-dsl-mastery/src/graph-manager/DSL_Mastery_GraphManagerHelper.ts:115]: done, corrected the error messages

File: packages/legend-extension-dsl-mastery/src/graph-manager/action/changeDetection/DSL_DataProvider_ObserverHelper.ts [hunk 2]
```diff
@@ -0,0 +1,36 @@
+/**
+ * Copyright (c) 2020-present, Goldman Sachs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
```
[the remark this answers is not in the document]
Comment (arikpamnani) [reply packages/legend-extension-dsl-mastery/src/graph-manager/action/changeDetection/DSL_DataProvider_ObserverHelper.ts]: renamed to have prefix `DSL_Mastery`

File: packages/legend-extension-dsl-mastery/src/graph-manager/protocol/pure/DSL_Mastery_PureProtocolProcessorPlugin.ts [hunk 3]
```diff
@@ -68,6 +110,90 @@ export class DSL_Mastery_PureProtocolProcessorPlugin
 
   override V1_getExtraElementBuilders(): V1_ElementBuilder<V1_PackageableElement>[] {
     return [
+      new V1_ElementBuilder<V1_KafkaConnection>({
+        elementClassName: 'KafkaConnection',
```
[the remark this answers is not in the document]
Comment (arikpamnani) [reply packages/legend-extension-dsl-mastery/src/graph-manager/protocol/pure/DSL_Mastery_PureProtocolProcessorPlugin.ts]: added a constant

File: packages/legend-extension-dsl-mastery/src/graph-manager/protocol/pure/DSL_Mastery_PureProtocolProcessorPlugin.ts [hunk 4]
```diff
@@ -68,6 +110,90 @@ export class DSL_Mastery_PureProtocolProcessorPlugin
...
+        secondPass: (
+          elementProtocol: V1_PackageableElement,
+          context: V1_GraphBuilderContext,
+        ): void => {
+          assertType(elementProtocol, V1_KafkaConnection);
+          V1_buildKafkaConnection(elementProtocol, context);
+        },
+      }),
+      new V1_ElementBuilder<V1_FTPConnection>({
+        elementClassName: 'FTPConnection',
```
[the remark this answers is not in the document]
Comment (arikpamnani) [reply packages/legend-extension-dsl-mastery/src/graph-manager/protocol/pure/DSL_Mastery_PureProtocolProcessorPlugin.ts]: added a constant

File: packages/legend-extension-dsl-mastery/src/graph-manager/protocol/pure/DSL_Mastery_PureProtocolProcessorPlugin.ts [hunk 5]
```diff
@@ -68,6 +110,90 @@ export class DSL_Mastery_PureProtocolProcessorPlugin
...
+        secondPass: (
+          elementProtocol: V1_PackageableElement,
+          context: V1_GraphBuilderContext,
+        ): void => {
+          assertType(elementProtocol, V1_FTPConnection);
+          V1_buildFTPConnection(elementProtocol, context);
+        },
+      }),
+      new V1_ElementBuilder<V1_HTTPConnection>({
+        elementClassName: 'HTTPConnection',
```
[the remark this answers is not in the document]
Comment (arikpamnani) [reply packages/legend-extension-dsl-mastery/src/graph-manager/protocol/pure/DSL_Mastery_PureProtocolProcessorPlugin.ts]: added a constant

File: packages/legend-extension-dsl-mastery/src/graph-manager/protocol/pure/DSL_Mastery_PureProtocolProcessorPlugin.ts [hunk 6]
```diff
@@ -96,6 +222,34 @@ export class DSL_Mastery_PureProtocolProcessorPlugin
           V1_buildMasterRecordDefinition(elementProtocol, context);
         },
       }),
+      new V1_ElementBuilder<V1_DataProvider>({
+        elementClassName: 'DataProvider',
```
[the remark this answers is not in the document]
Comment (arikpamnani) [reply packages/legend-extension-dsl-mastery/src/graph-manager/protocol/pure/DSL_Mastery_PureProtocolProcessorPlugin.ts]: added a constant

File: packages/legend-extension-dsl-mastery/src/graph-manager/protocol/pure/v1/transformation/pureGraph/to/V1_DSL_Mastery_BuilderHelper.ts [hunk 7]
```diff
@@ -69,20 +139,371 @@ import {
...
+  context: V1_GraphBuilderContext,
+): Trigger => {
+  switch (element.constructor) {
+    case V1_ManualTrigger:
+      return new ManualTrigger();
+    case V1_CronTrigger:
+      return V1_buildCronTrigger(element as V1_CronTrigger, context);
+    default:
+      throw new UnsupportedOperationError(
+        `Unsupported trigger '${typeof element}'`,
```
[the remark this answers is not in the document]
Comment (arikpamnani) [reply packages/legend-extension-dsl-mastery/src/graph-manager/protocol/pure/v1/transformation/pureGraph/to/V1_DSL_Mastery_BuilderHelper.ts]: done, updated the error messages

File: packages/legend-extension-dsl-mastery/src/graph-manager/action/changeDetection/DSL_Connection_ObserverHelper.ts [hunk 8]
```diff
@@ -0,0 +1,68 @@
+/**
+ * Copyright (c) 2020-present, Goldman Sachs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
```
[the remark this answers is not in the document]
Comment (arikpamnani) [reply packages/legend-extension-dsl-mastery/src/graph-manager/action/changeDetection/DSL_Connection_ObserverHelper.ts]: renamed to have prefix `DSL_Mastery`

File: packages/legend-extension-dsl-mastery/src/graph-manager/protocol/pure/v1/model/packageableElements/mastery/V1_DSL_Mastery_AcquisitionProtocol.ts [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+  }
+}
+
+export class V1_FileAcquisitionProtocol extends V1_AcquisitionProtocol {
+  connection!: string;
+  filePath!: string;
+  fileType!: V1_FileType;
+  fileSplittingKeys: string[] | undefined;
+  headerLines!: number;
+  recordsKey: string | undefined;
```
[the remark this answers is not in the document]
Comment (arikpamnani) [reply packages/legend-extension-dsl-mastery/src/graph-manager/protocol/pure/v1/model/packageableElements/mastery/V1_DSL_Mastery_AcquisitionProtocol.ts]: done

File: packages/legend-extension-dsl-mastery/src/graph-manager/protocol/pure/v1/model/packageableElements/mastery/V1_DSL_Mastery_AcquisitionProtocol.ts [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+  override get hashCode(): string {
+    return hashArray([
+      MASTERY_HASH_STRUCTURE.REST_SERVICE_ACQUISITION_PROTOCOL,
+      super.hashCode,
+    ]);
+  }
+}
+
+export enum V1_FileType {
+  JSON,
```
[the remark this answers is not in the document]
Comment (arikpamnani) [reply packages/legend-extension-dsl-mastery/src/graph-manager/protocol/pure/v1/model/packageableElements/mastery/V1_DSL_Mastery_AcquisitionProtocol.ts]: changed to defined at the start

File: packages/legend-extension-dsl-mastery/src/graph-manager/protocol/pure/v1/model/packageableElements/mastery/V1_DSL_Mastery_DataProvider.ts [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+
+  accept_PackageableElementVisitor<T>(
+    visitor: V1_PackageableElementVisitor<T>,
+  ): T {
+    return visitor.visit_PackageableElement(this);
+  }
+}
+
+export enum V1_DataProviderType {
+  AGGREGATOR = 'Aggregator',
```
[the remark this answers is not in the document]
Comment (arikpamnani) [reply packages/legend-extension-dsl-mastery/src/graph-manager/protocol/pure/v1/model/packageableElements/mastery/V1_DSL_Mastery_DataProvider.ts]: changed to defined at the start

File: packages/legend-extension-dsl-mastery/src/graph-manager/protocol/pure/v1/model/packageableElements/mastery/V1_DSL_Mastery_Trigger.ts [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+  July,
+  August,
+  September,
+  October,
+  November,
+  December,
+}
+
+export enum V1_Day {
+  Monday,
```
[the remark this answers is not in the document]
Comment (arikpamnani) [reply packages/legend-extension-dsl-mastery/src/graph-manager/protocol/pure/v1/model/packageableElements/mastery/V1_DSL_Mastery_Trigger.ts]: changed to defined at the start

File: packages/legend-extension-dsl-mastery/src/graph-manager/protocol/pure/v1/transformation/pureGraph/from/V1_DSL_Mastery_TransformerHelper.ts [hunk 13]
```diff
@@ -64,26 +111,374 @@ import {
...
+  context: V1_GraphTransformerContext,
+): V1_Trigger => {
+  switch (element.constructor) {
+    case ManualTrigger:
+      return new V1_ManualTrigger();
+    case CronTrigger:
+      return V1_transformCronTrigger(element as CronTrigger, context);
+    default:
+      throw new UnsupportedOperationError(
+        `Unsupported trigger '${typeof element}'`,
```
[the remark this answers is not in the document]
Comment (arikpamnani) [reply packages/legend-extension-dsl-mastery/src/graph-manager/protocol/pure/v1/transformation/pureGraph/from/V1_DSL_Mastery_TransformerHelper.ts]: done, updated the error messages

File: packages/legend-extension-dsl-mastery/src/graph-manager/protocol/pure/v1/transformation/pureProtocol/V1_DSL_Mastery_ProtocolHelper.ts [hunk 14]
```diff
@@ -50,45 +50,529 @@ import {
...
+  });
+
+export const V1_TokenAuthenticationStrategyProtocolSchema = (
+  plugins: PureProtocolProcessorPlugin[],
+): ModelSchema<V1_TokenAuthenticationStrategy> =>
+  createModelSchema(V1_TokenAuthenticationStrategy, {
+    _type: usingConstantValueSchema(V1_AuthenticationStrategyType.TOKEN),
+    credential: optional(
+      custom(
+        (val) => (val ? V1_serializeCredentialSecret(val, plugins) : SKIP),
```
[the remark this answers is not in the document]
Comment (arikpamnani) [reply packages/legend-extension-dsl-mastery/src/graph-manager/protocol/pure/v1/transformation/pureProtocol/V1_DSL_Mastery_ProtocolHelper.ts]: changed to use `optionalCustom`

File: packages/legend-extension-dsl-mastery/src/graph-manager/protocol/pure/v1/transformation/pureProtocol/V1_DSL_Mastery_ProtocolHelper.ts [hunk 15]
```diff
@@ -50,45 +50,529 @@ import {
...
+});
+
+export const V1_manualTriggerSchema = createModelSchema(V1_ManualTrigger, {
+  _type: usingConstantValueSchema(V1_TriggerType.MANUAL),
+});
+
+export const V1_serializeTrigger = (
+  protocol: V1_Trigger,
+): PlainObject<V1_Trigger> => {
+  switch (protocol.constructor) {
```
[the remark this answers is not in the document]
Comment (arikpamnani) [reply packages/legend-extension-dsl-mastery/src/graph-manager/protocol/pure/v1/transformation/pureProtocol/V1_DSL_Mastery_ProtocolHelper.ts]: updated to `instanceOf`

File: packages/legend-extension-dsl-mastery/src/graph-manager/protocol/pure/v1/transformation/pureGraph/from/V1_DSL_Mastery_TransformerHelper.ts [hunk 16]
```diff
@@ -64,26 +111,374 @@ import {
...
+
+/**********
+ * authorization
+ **********/
+
+export const V1_transformAuthorization = (
+  element: Authorization | undefined,
+  context: V1_GraphTransformerContext,
+): V1_Authorization | undefined => {
+  if (element === undefined) {
```
[the remark this answers is not in the document]
Comment (arikpamnani) [reply packages/legend-extension-dsl-mastery/src/graph-manager/protocol/pure/v1/transformation/pureGraph/from/V1_DSL_Mastery_TransformerHelper.ts]: updated methods to not take `undefined` and handled `undefined` during invocation

File: packages/legend-extension-dsl-mastery/src/graph-manager/protocol/pure/v1/transformation/pureGraph/from/V1_DSL_Mastery_TransformerHelper.ts [hunk 17]
```diff
@@ -64,26 +111,374 @@ import {
...
+export const V1_transformRestAcquisitionProtocol = (
+  element: RestAcquisitionProtocol,
+  context: V1_GraphTransformerContext,
+): V1_RestAcquisitionProtocol => new V1_RestAcquisitionProtocol();
+
+export const V1_transformAcquisitionProtocol = (
+  element: AcquisitionProtocol,
+  context: V1_GraphTransformerContext,
+): V1_AcquisitionProtocol => {
+  switch (element.constructor) {
```
[the remark this answers is not in the document]
Comment (arikpamnani) [reply packages/legend-extension-dsl-mastery/src/graph-manager/protocol/pure/v1/transformation/pureGraph/from/V1_DSL_Mastery_TransformerHelper.ts]: done, using instanceof checks on models

File: packages/legend-extension-dsl-mastery/src/graph-manager/protocol/pure/v1/transformation/pureGraph/from/V1_DSL_Mastery_TransformerHelper.ts [hunk 18]
```diff
@@ -64,26 +111,374 @@ import {
...
+};
+
+export const V1_transformAuthenticationStrategy = (
+  element: AuthenticationStrategy | undefined,
+  context: V1_GraphTransformerContext,
+): V1_AuthenticationStrategy | undefined => {
+  if (element === undefined) {
+    return element;
+  }
+  switch (element.constructor) {
```
[the remark this answers is not in the document]
Comment (arikpamnani) [reply packages/legend-extension-dsl-mastery/src/graph-manager/protocol/pure/v1/transformation/pureGraph/from/V1_DSL_Mastery_TransformerHelper.ts]: done, using instanceof checks on models

File: packages/legend-extension-dsl-mastery/src/graph-manager/protocol/pure/v1/transformation/pureGraph/from/V1_DSL_Mastery_TransformerHelper.ts [hunk 19]
```diff
@@ -64,26 +111,374 @@ import {
...
+
+export const V1_transformDataProvider = (
+  element: DataProvider,
+  context: V1_GraphTransformerContext,
+): V1_DataProvider => {
+  const dataProvider = new V1_DataProvider();
+  V1_initPackageableElement(dataProvider, element);
+  dataProvider.dataProviderId = element.dataProviderId;
+  dataProvider.dataProviderType =
+    element.dataProviderType.valueOf() as V1_DataProviderType;
```
[the remark this answers is not in the document]
Comment (arikpamnani) [reply packages/legend-extension-dsl-mastery/src/graph-manager/protocol/pure/v1/transformation/pureGraph/from/V1_DSL_Mastery_TransformerHelper.ts:152]: We could do that, but isn't the idea to keep protocols and metamodels completely separate (not dependent on each other or have any common dependency)? I had checked the persistence DSL for reference, e.g. -

- https://github.com/arikpamnani/legend-studio/blob/b2e14b15379eef36e39d906d315fd4fb96472cd6/packages/legend-extension-dsl-persistence/src/graph/metamodel/pure/model/packageableElements/persistence/DSL_Persistence_TargetShape.ts#L75
- https://github.com/arikpamnani/legend-studio/blob/4e7b750ee649033b66c87b84b4ff242ad3829580/packages/legend-extension-dsl-persistence/src/graph-manager/protocol/pure/v1/model/packageableElements/persistence/V1_DSL_Persistence_TargetShape.ts#L74

**Review by MauricioUyaguari — APPROVED** [review]

**Pull request merged**